rts/lra/lra-proxy/api/src/main/java/io/narayana/lra/client/internal/proxy/nonjaxrs/LRACDIExtension.java

```diff
@@ -0,0 +1,150 @@
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package io.narayana.lra.client.internal.proxy.nonjaxrs;
+
+import org.eclipse.microprofile.lra.annotation.Compensate;
+import org.eclipse.microprofile.lra.annotation.Complete;
+import org.eclipse.microprofile.lra.annotation.Forget;
+import org.eclipse.microprofile.lra.annotation.ParticipantStatus;
+import org.eclipse.microprofile.lra.annotation.Status;
+import org.eclipse.microprofile.lra.annotation.ws.rs.LRA;
+import org.eclipse.microprofile.lra.participant.InvalidLRAParticipantDefinitionException;
+
+import javax.enterprise.event.Observes;
+import javax.enterprise.inject.spi.Extension;
+import javax.enterprise.inject.spi.ProcessAnnotatedType;
+import javax.enterprise.inject.spi.WithAnnotations;
+import javax.ws.rs.Path;
+import javax.ws.rs.core.Response;
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Method;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.concurrent.CompletionStage;
+
+public class LRACDIExtension implements Extension {
+
+    private String currentMethodGenericString;
+
+    private LRAParticipantRegistry participantRegistry = LRAParticipantRegistry.getInstance();
+
+    public void observe(@Observes @WithAnnotations(Path.class) ProcessAnnotatedType<?> type) {
+        Class<?> javaClass = type.getAnnotatedType().getJavaClass();
+
+        if (isNotLRAParticipant(javaClass)) {
+            return;
+        }
+
+        LRAParticipant participant = new LRAParticipant(javaClass);
+
+        Arrays.stream(javaClass.getDeclaredMethods()).forEach(m -> processParticipantMethod(m, participant));
+
+        boolean shouldRegister = participant.getCompensateMethod() != null ||
+            participant.getCompleteMethod() != null ||
+            participant.getStatusMethod() != null ||
+            participant.getForgetMethod() != null;
+
+        if (shouldRegister) {
+            participantRegistry.registerParticipant(participant);
+        }
+    }
+
+    private boolean isNotLRAParticipant(Class<?> javaClass) {
+        return Arrays.stream(javaClass.getDeclaredMethods()).noneMatch(m ->
+            Arrays.stream(m.getDeclaredAnnotations()).anyMatch(a -> a.annotationType().equals(LRA.class)));
+    }
+
+    /**
+     * Process participant method for non JAX-RS related processing
+     * defined by the specification and verify its signature
+     * @param method method to be processed
+     *
+     * @throws InvalidLRAParticipantDefinitionException if error is detected
+     */
+    private void processParticipantMethod(Method method, LRAParticipant participant) {
+
+        if (isJaxRsMethod(method) || isNotParticipantMethod(method)) {
```

Comment:
Related to this is one of my comments on your update to this line:

> You have already (effectively) performed this set of tests when you called getParticipantAnnotation on line 93. You should set the method as soon as you discover the annotation type.

So I will leave this conversation as unresolved for now since it is related.